src/main/scala/mesosphere/marathon/core/storage/simple/PersistenceStore.scala

```diff
@@ -0,0 +1,124 @@
+    * created if they don't already exist. If there is already a node with the same path a [[org.apache.zookeeper.KeeperException.NodeExistsException]]
+    * is thrown.
+    *
+    * @return
+    */
+  def create: Flow[Node, String, NotUsed]
+  def create(node: Node): Source[String, NotUsed] = Source.single(node).via(create)
+  def create(nodes: Seq[Node]): Source[String, NotUsed] = Source.fromIterator(() => nodes.iterator).via(create)
+
+  /**
+    * A Flow for reading nodes from the store. It takes a stream of node paths and returns a stream of Try[Node] elements.
+    * It's a Success[Node] for existing nodes or Failure(e) for non-existing nodes where e is an instance of
+    * [[org.apache.zookeeper.KeeperException.NoNodeException]]. The exception contains the path to the node which
+    * simplifies handling failed reads. For other exceptions stream is completed with a failure.
+    *
+    * @return
+    */
+  def read: Flow[String, Try[Node], NotUsed]
+  def read(path: String): Source[Try[Node], NotUsed] = Source.single(path).via(read)
+  def read(paths: Seq[String]): Source[Try[Node], NotUsed] = Source.fromIterator(() => paths.iterator).via(read)
+
+  /**
+    * A Flow for updating nodes in the store. It takes a stream of nodes and returns a stream of paths to indicate
+    * a successful update operation for the returned path. Only existing nodes can be updated. If a node with the path
+    * does not exist a [[org.apache.zookeeper.KeeperException.NoNodeException]] is thrown.
+    *
+    * @return
+    */
+  def update: Flow[Node, String, NotUsed]
+  def update(node: Node): Source[String, NotUsed] = Source.single(node).via(update)
+  def update(nodes: Seq[Node]): Source[String, NotUsed] = Source.fromIterator(() => nodes.iterator).via(update)
+
+  /**
+    * A Flow for deleting nodes from the repository. It takes a stream of paths and returns a stream of paths to indicate
+    * a successful deletion operation for the returned path. If the node doesn't exist the operation is still considered
+    * successful.
+    *
+    * @return
+    */
+  def delete: Flow[String, String, NotUsed]
+  def delete(path: String): Source[String, NotUsed] = Source.single(path).via(delete)
+  def delete(paths: Seq[String]): Source[String, NotUsed] = Source.fromIterator(() => paths.iterator).via(delete)
+
+  /**
+    * Returns the list of paths for children nodes for the passed path. Note that returned path is relative and does
+    * not contain node's path prefix.
+    *
+    * @param path parent node's path
+    * @return
+    */
+  def children(path: String): Future[Seq[String]]
+
+  /**
+    * Checks for the existence of a node with passed path.
+    *
+    * @param path node's path
+    * @return
+    */
+  def exists(path: String): Future[Boolean]
+
+  /**
+    * Method syncs state with underlying store.
+    *
+    * @param path node path to sync
+    * @return
+    */
+  def sync(path: String): Future[Done]
```

Comment:
So, when is a Node actually persisted?

Reply:
Not sure I understand what you mean. If you look at the implementation, then calling [forPath](https://github.com/mesosphere/marathon/pull/6265/files#diff-68cf4611e0b949cf28991469f497ae2fR68) on the builder submits the request to the curator.